File: screen-switcher/src/main/java/com/jaynewstrom/screenswitcher/ScreenSwitcherFactory.java
package com.jaynewstrom.screenswitcher;

import android.app.Activity;
import android.view.View;
import android.view.ViewGroup;
import android.view.ViewGroup.LayoutParams;

import static com.jaynewstrom.screenswitcher.Preconditions.checkArgument;
import static com.jaynewstrom.screenswitcher.Preconditions.checkNotNull;

public final class ScreenSwitcherFactory {

    private ScreenSwitcherFactory() {
    }

    /**
     * Creates a {@link ScreenSwitcher} that uses {@link Activity#addContentView(View, LayoutParams)} to display a {@link Screen}.
     */
    public static ScreenSwitcher activityScreenSwitcher(Activity activity, ScreenSwitcherState state,
            ScreenSwitcherPopHandler popHandler) {
        checkNotNull(activity, "activity == null");
        validateState(state);
        checkNotNull(popHandler, "popHandler == null");
        return new RealScreenSwitcher(activity, state, new ActivityScreenSwitcherHost(activity, popHandler));
    }

    /**
     * Creates a {@link ScreenSwitcher} that uses {@link ViewGroup#addView(View)} to display a {@link Screen}.
     */
    public static ScreenSwitcher viewScreenSwitcher(ViewGroup viewGroup, ScreenSwitcherState state,
            ScreenSwitcherPopHandler popHandler) {
        checkNotNull(viewGroup, "viewGroup == null");
        checkNotNull(popHandler, "popHandler == null");
        validateState(state);
        return new RealScreenSwitcher(viewGroup.getContext(), state, new ViewScreenSwitcherHost(viewGroup, popHandler));
    }

    private static void validateState(ScreenSwitcherState state) {
        checkNotNull(state, "state == null");
        checkArgument(state.getScreens().size() >= 1, "state needs screens in order to initialize a ScreenSwitcher");
    }
}

File: screen-switcher/src/main/java/com/jaynewstrom/screenswitcher/ScreenSwitcherFactory.kt
package com.jaynewstrom.screenswitcher

import android.app.Activity
import android.view.ViewGroup

object ScreenSwitcherFactory {
    /**
     * Creates a [ScreenSwitcher] that uses [Activity.addContentView] to display a [Screen].
     */
    @JvmStatic fun activityScreenSwitcher(
        activity: Activity,
        state: ScreenSwitcherState,
        finishHandler: ScreenSwitcherFinishHandler
    ): ScreenSwitcher {
        return RealScreenSwitcher(state, ActivityScreenSwitcherHost(activity, finishHandler))
    }

    /**
     * Creates a [ScreenSwitcher] that uses [ViewGroup.addView] to display a [Screen].
     */
    @JvmStatic fun viewScreenSwitcher(
        viewGroup: ViewGroup,
        state: ScreenSwitcherState,
        finishHandler: ScreenSwitcherFinishHandler
    ): ScreenSwitcher {
        return RealScreenSwitcher(state, ViewScreenSwitcherHost(viewGroup, finishHandler))
    }
}
